### DISCUSSION TITLE
Tauri 开发 FAQ

### DISCUSSION THREAD

**Discussion by lencx** [lencx/tauri-tutorial#9] [Blog]
### tauri build error

`Cargo.toml` 中的 `tauri`，`tauri-build` 版本需要与 `package.json` 中的 `@tauri-apps/cli`，`@tauri-apps/api` 版本保持一致。

```bash
error: failed to run custom build command for `omb v0.1.0 (/Users/lencx/github/OhMyBox/src-tauri)`

Caused by:
  process didn't exit successfully: `/Users/lencx/github/OhMyBox/src-tauri/target/debug/build/omb-4ae2dd3579447390/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=TAURI_CONFIG
  cargo:rerun-if-changed=tauri.conf.json

  --- stderr
  thread 'main' panicked at 'error found during tauri-build: Error("invalid type: null, expected a boolean", line: 1, column: 944)', /Users/lencx/.cargo/registry/src/github.com-1ecc6299db9ec823/tauri-build-1.0.0/src/lib.rs:172:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
🚀 A new version of Tauri CLI is avaliable! [1.0.0]
error Command failed with exit code 101.
```

### error: failed to run custom build command for `webview2-com-sys v0.18.0`

请按照 Tauri 提供的 [Windows 安装指南](https://tauri.app/v1/guides/getting-started/prerequisites/#setting-up-windows)进行操作。如果仍无法解决，请查看相关 issues：

- [[bug] error: failed to run custom build command for webview2-com-sys v0.13.0 in windows](https://github.com/tauri-apps/tauri/issues/4124)
- [failed to run custom build command for webview2-com-sys v0.13.0](https://github.com/tauri-apps/tauri/issues/4189)

### [vite] warning: Top-level "this" will be replaced with undefined since this file is an ECMAScript module

这是因为 `plugin-react` 使用 `@babel/plugin-transform-react-jsx` 来转换 JSX。 可以设置 esbuild.logOverride 以忽略警告。

```bash
# https://github.com/vitejs/vite/issues/8644#issuecomment-1159308803
warning: `omb` (bin "omb") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 6.72s
     Running `target/debug/omb`
4:43:49 PM [vite] warning: Top-level "this" will be replaced with undefined since this file is an ECMAScript module
16 |      lineNumber: 13,
17 |      columnNumber: 14
18 |    }, this)
   |       ^
19 |  }, {
20 |    path: 'tools/canvas',

  Plugin: vite:esbuild
  File: /Users/lencx/github/OhMyBox/src/routes.tsx
```

```js
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
  esbuild: {
    logOverride: { 'this-is-undefined-in-esm': 'silent' }
  }
})
```

### error: linking with `cc` failed: exit status: 1

macOS 系统升级导致此错误，可以通过 `xcode-select --install` 来解决。

```bash
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "-arch" "x86_64" ...
...
  = note: xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun
```

### xcode-select: note: install requested for command line developer tools

更新 Xcode 命令行工具最简单方法是删除旧版本，然后安装新版本。

```bash
sudo rm -rf /Library/Developer/CommandLineTools
xcode-select --install
```

### Xcode Error When Launching Terminal

每次升级后 XCode 都会发生这种情况：

```bash
2022-07-24 07:54:12.487 xcodebuild[1657:2298853] Requested but did not find extension point with identifier Xcode.IDEKit.ExtensionSentinelHostApplications for extension Xcode.DebuggerFoundation.AppExtensionHosts.watchOS of plug-in com.apple.dt.IDEWatchSupportCore
2022-07-24 07:54:12.487 xcodebuild[1657:2298853] Requested but did not find extension point with identifier Xcode.IDEKit.ExtensionPointIdentifierToBundleIdentifier for extension Xcode.DebuggerFoundation.AppExtensionToBundleIdentifierMap.watchOS of plug-in com.apple.dt.IDEWatchSupportCore
```

解决办法，重装 Xcode 命令行工具：

```bash
sudo rm -rf /Library/Developer/CommandLineTools
xcode-select --install
```

### 插件冲突

1. [Rust](https://marketplace.visualstudio.com/publishers/rust-lang) - The Rust Programming Language
2. [rust-analyzer](https://marketplace.visualstudio.com/items?itemName=matklad.rust-analyzer) - An alternative rust language server to the RLS

同时安装 1，2 两个插件则会报以下错误，关闭 1 插件即可解决。

```
You have both the rust-analyzer (matklad.rust-analyzer) and Rust (rust-lang.rust) plugins enabled.
These are known to conflict and cause various functions of both plugins to not work correctly.
You should disable one of them.You have both the rust-analyzer (matklad.rust-analyzer) and 
Rust (rust-lang.rust) plugins enabled. These are known to conflict and cause various functions 
of both plugins to not work correctly. You should disable one of them.
```

### Couldn’t start client Rust Language Server

vscode 无法找到 `rustup` 可执行文件的路径，可以通过手动配置 `settings.json` 来解决。

```json
"rust-client.rustupPath": "$HOME/.cargo/bin/rustup"
```

---

## 相关阅读

- [[vite] warning: Top-level "this" will be replaced with undefined since this file is an ECMAScript module](https://github.com/vitejs/vite/issues/8644)
- [How do I solve "Couldn't start client Rust Language Server" with the Rust VS Code extension?](https://stackoverflow.com/questions/60816886/how-do-i-solve-couldnt-start-client-rust-language-server-with-the-rust-vs-cod)
- [How to update Xcode from command line](https://stackoverflow.com/questions/34617452/how-to-update-xcode-from-command-line)
- [Xcode Error When Launching Terminal](https://apple.stackexchange.com/questions/438785/xcode-error-when-launching-terminal)

**Comment by ljybill** [comment]
tauri可以注入js到webview吗？是否可以做个示例

**Reply by lencx** [reply]
Tauri 主要分两部分：
- 后端 - Rust 实现，提供和系统通信的能力
- 前端 - 借助 web 生态，使用 webview 来渲染页面

注入 js 是什么意思，而且 tauri 本身也很注重安全，设置了各种访问权限，我不太理解你所说的注入是否属于骚操作？

**Reply by ljybill** [reply]
有点类似浏览器插件，比如我通过tauri打开了baidu站点，然后我通过注入一段js代码（比如 `console.log('Hello')`），那么在控制台上能看到log输出。这种效果

**Reply by lencx** [reply]
tauri 是打包成应用形式的，这样注入的意义是什么？你要是想实现插件机制，可以在开发时自己去支持，我理解和 tauri 没啥关系吧。

**Reply by lencx** [reply]
现在重新回答一下你的问题：通过 tauri rust 中的 `tauri::WindowBuilder::new` 新开的窗口可以通过 `initialization_script` 来注入外部脚本。

**Comment by yorunning** [comment]
调用服务器api接口应该放在前端做还在通过tauricore来调用呢
